In [391]:
import pandas as pd

In [392]:
resume_data = pd.read_csv("data\\Resume_data_second\\resume_data.csv")

In [393]:
resume_data.head(5)

address                                   career_objective  \
0     NaN  Big data analytics working and database wareho...   
1     NaN  Fresher looking to join as a data analyst and ...   
2     NaN                                                NaN   
3     NaN  To obtain a position in a fast-paced business ...   
4     NaN  Professional accountant with an outstanding wo...   

                                              skills  \
0  ['Big Data', 'Hadoop', 'Hive', 'Python', 'Mapr...   
1  ['Data Analysis', 'Data Analytics', 'Business ...   
2  ['Software Development', 'Machine Learning', '...   
3  ['accounts payables', 'accounts receivables', ...   
4  ['Analytical reasoning', 'Compliance testing k...   

                        educational_institution_name  \
0  ['The Amity School of Engineering & Technology...   
1  ['Delhi University - Hansraj College', 'Delhi ...   
2    ['Birla Institute of Technology (BIT), Ranchi']   
3  ['Martinez Adult Education, Business Training ...   
4                          ['Kent State University']   

                                        degree_names     passing_years  \
0                                         ['B.Tech']          ['2019']   
1    ['B.Sc (Maths)', 'M.Sc (Science) (Statistics)']  ['2015', '2018']   
2                                         ['B.Tech']          ['2018']   
3  ['Computer Applications Specialist Certificate...          ['2008']   
4            ['Bachelor of Business Administration']            [None]   

  educational_results    result_types             major_field_of_studies  \
0             ['N/A']          [None]                    ['Electronics']   
1      ['N/A', 'N/A']  ['N/A', 'N/A']      ['Mathematics', 'Statistics']   
2             ['N/A']         ['N/A']  ['Electronics/Telecommunication']   
3              [None]          [None]          ['Computer Applications']   
4            ['3.84']          [None]                     ['Accounting']   

                          professional_company_names  ... online_links  \
0                                      ['Coca-COla']  ...          NaN   
1                                ['BIB Consultancy']  ...          NaN   
2                              ['Axis Bank Limited']  ...          NaN   
3  ['Company Name ï¼ City , State', 'Company Name...  ...          NaN   
4  ['Company Name', 'Company Name', 'Company Name...  ...       [None]   

  issue_dates           expiry_dates  \
0         NaN                    NaN   
1         NaN                    NaN   
2         NaN                    NaN   
3         NaN                    NaN   
4      [None]  ['February 15, 2021']   

                                  job_position_name  \
0                           Senior Software Engineer   
1                     Machine Learning (ML) Engineer   
2  Executive/ Senior Executive- Trade Marketing, ...   
3                     Business Development Executive   
4                                Senior iOS Engineer   

                            educationaL_requirements experiencere_requirement  \
0  B.Sc in Computer Science & Engineering from a ...          At least 1 year   
1  M.Sc in Computer Science & Engineering or in a...       At least 5 year(s)   
2            Master of Business Administration (MBA)         At least 3 years   
3                                    Bachelor/Honors             1 to 3 years   
4      Bachelor of Science (BSc) in Computer Science         At least 4 years   

      age_requirement                                 responsibilities.1  \
0                 NaN  Technical Support\nTroubleshooting\nCollaborat...   
1                 NaN  Machine Learning Leadership\nCross-Functional ...   
2                 NaN  Trade Marketing Executive\nBrand Visibility, S...   
3  Age 22 to 30 years  Apparel Sourcing\nQuality Garment Sourcing\nRe...   
4                 NaN  iOS Lifecycle\nRequirement Analysis\nNative Fr...   

                                     skills_required matched_score  
0    

In [394]:
resume_data.columns

Index(['address', 'career_objective', 'skills', 'educational_institution_name',
       'degree_names', 'passing_years', 'educational_results', 'result_types',
       'major_field_of_studies', 'professional_company_names', 'company_urls',
       'start_dates', 'end_dates', 'related_skils_in_job', 'positions',
       'locations', 'responsibilities', 'extra_curricular_activity_types',
       'extra_curricular_organization_names',
       'extra_curricular_organization_links', 'role_positions', 'languages',
       'proficiency_levels', 'certification_providers', 'certification_skills',
       'online_links', 'issue_dates', 'expiry_dates', 'job_position_name',
       'educationaL_requirements', 'experiencere_requirement',
       'age_requirement', 'responsibilities.1', 'skills_required',
       'matched_score'],
      dtype='object')

In [395]:
for i in resume_data.columns:
  print( i , " is ", resume_data[i][0])  

address  is  nan
career_objective  is  Big data analytics working and database warehouse manager with robust experience in handling all kinds of data. I have also used multiple cloud infrastructure services and am well acquainted with them. Currently in search of role that offers more of development.
skills  is  ['Big Data', 'Hadoop', 'Hive', 'Python', 'Mapreduce', 'Spark', 'Java', 'Machine Learning', 'Cloud', 'Hdfs', 'YARN', 'Core Java', 'Data Science', 'C++', 'Data Structures', 'DBMS', 'RDBMS', 'Informatica', 'Talend', 'Amazon Redshift', 'Microsoft Azure']
educational_institution_name  is  ['The Amity School of Engineering & Technology (ASET), Noida']
degree_names  is  ['B.Tech']
passing_years  is  ['2019']
educational_results  is  ['N/A']
result_types  is  [None]
major_field_of_studies  is  ['Electronics']
professional_company_names  is  ['Coca-COla']
company_urls  is  [None]
start_dates  is  ['Nov 2019']
end_dates  is  ['Till Date']
related_skils_in_job  is  [['Big Data']]
position

In [396]:
print(resume_data.shape)

(9544, 35)


In [397]:
resume_data = resume_data.drop(columns=['passing_years','professional_company_names', 'company_urls',
       'locations','certification_skills',
       'online_links', 'issue_dates', 'expiry_dates', 'job_position_name',
       'educationaL_requirements', 'experiencere_requirement',
       'age_requirement', 'responsibilities.1', 'skills_required',
       'matched_score'])

In [398]:
print(resume_data.shape)
resume_data.head(1)

(9544, 20)


address                                   career_objective  \
0     NaN  Big data analytics working and database wareho...   

                                              skills  \
0  ['Big Data', 'Hadoop', 'Hive', 'Python', 'Mapr...   

                        educational_institution_name degree_names  \
0  ['The Amity School of Engineering & Technology...   ['B.Tech']   

  educational_results result_types major_field_of_studies   start_dates  \
0             ['N/A']       [None]        ['Electronics']  ['Nov 2019']   

       end_dates related_skils_in_job             positions  \
0  ['Till Date']       [['Big Data']]  ['Big Data Analyst']   

                                    responsibilities  \
0  Technical Support\nTroubleshooting\nCollaborat...   

  extra_curricular_activity_types extra_curricular_organization_names  \
0                             NaN                                 NaN   

  extra_curricular_organization_links role_positions languages  \
0                                 NaN            NaN       NaN   

  proficiency_levels certification_providers  
0                NaN                     NaN

In [399]:
resume_data = resume_data.drop_duplicates()
print(resume_data.shape)
resume_data.head(4)

(9544, 20)


address                                   career_objective  \
0     NaN  Big data analytics working and database wareho...   
1     NaN  Fresher looking to join as a data analyst and ...   
2     NaN                                                NaN   
3     NaN  To obtain a position in a fast-paced business ...   

                                              skills  \
0  ['Big Data', 'Hadoop', 'Hive', 'Python', 'Mapr...   
1  ['Data Analysis', 'Data Analytics', 'Business ...   
2  ['Software Development', 'Machine Learning', '...   
3  ['accounts payables', 'accounts receivables', ...   

                        educational_institution_name  \
0  ['The Amity School of Engineering & Technology...   
1  ['Delhi University - Hansraj College', 'Delhi ...   
2    ['Birla Institute of Technology (BIT), Ranchi']   
3  ['Martinez Adult Education, Business Training ...   

                                        degree_names educational_results  \
0                                         ['B.Tech']             ['N/A']   
1    ['B.Sc (Maths)', 'M.Sc (Science) (Statistics)']      ['N/A', 'N/A']   
2                                         ['B.Tech']             ['N/A']   
3  ['Computer Applications Specialist Certificate...              [None]   

     result_types             major_field_of_studies  \
0          [None]                    ['Electronics']   
1  ['N/A', 'N/A']      ['Mathematics', 'Statistics']   
2         ['N/A']  ['Electronics/Telecommunication']   
3          [None]          ['Computer Applications']   

                                         start_dates  \
0                                       ['Nov 2019']   
1                                       ['Sep 2019']   
2                                      ['June 2018']   
3  ['January 2011', 'January 2008', 'January 2006...   

                                           end_dates  \
0                                      ['Till Date']   
1                                      ['Till Date']   
2                                      ['Till Date']   
3  ['November 2015', 'January 2010', 'January 200...   

                                related_skils_in_job  \
0                                     [['Big Data']]   
1  [['Data Analysis', 'Business Analysis', 'Machi...   
2  [['Unified Payment Interface', 'Risk Predictio...   
3  [['accounts receivables', 'banking', 'G/L Acco...   

                                           positions  \
0                               ['Big Data Analyst']   
1                               ['Business Analyst']   
2  ['Software Developer (Machine Learning Enginee...   
3  ['Accountant', 'Accounts Receivable Clerk', 'M...   

                                    responsibilities  \
0  Technical Support\nTroubleshooting\nCollaborat...   
1  Machine Learning Leadership\nCross-Functional ...   
2  Trade Marketing Executive\nBrand Visibility, S...   
3  Apparel Sourcing\nQuality Garment Sourcing\nRe...   

  extra_curricular_activity_types extra_curricular_organization_names  \
0                             NaN                                 NaN   
1                             NaN                                 NaN   
2                             NaN                                 NaN   
3                             NaN                                 NaN   

  extra_curricular_organization_links role_positions languages  \
0                                 NaN            NaN       NaN   
1                                 NaN            NaN       NaN   
2                                 NaN            NaN       NaN   
3                                 NaN            NaN       NaN   

  proficiency_levels certification_providers  
0                NaN                     NaN  
1                NaN                     NaN  
2                NaN                     NaN  
3                NaN                     NaN

In [400]:
"""
i want to reach to new Colmns:-
  1- skill 
  2- location 
  3- qualification 
  4- experience year 
  5- work history 
  - so in first column i will merge columns in my Data ( skills , related_skills_in_job , language , proficiency_levels)
  - second column location i will take it from address column in my data 
  - third columnd qualification i will take it from (carrer objective, education_instuation_name, degree_name, major_field_of_studies )
  - fourth column Experience year i will take from ['start_dates', 'end_dates'] subtract together 
  - fifth column work_history i will take from columns ( position , Rule_position , responsibilities )
"""

"\ni want to reach to new Colmns:-\n  1- skill \n  2- location \n  3- qualification \n  4- experience year \n  5- work history \n  - so in first column i will merge columns in my Data ( skills , related_skills_in_job , language , proficiency_levels)\n  - second column location i will take it from address column in my data \n  - third columnd qualification i will take it from (carrer objective, education_instuation_name, degree_name, major_field_of_studies )\n  - fourth column Experience year i will take from ['start_dates', 'end_dates'] subtract together \n  - fifth column work_history i will take from columns ( position , Rule_position , responsibilities )\n"

#### First convert ( skills , related_skills_in_job , language , proficiency_levels) to skills column

In [401]:
print(resume_data.skills)
print(resume_data.related_skils_in_job)

0       ['Big Data', 'Hadoop', 'Hive', 'Python', 'Mapr...
1       ['Data Analysis', 'Data Analytics', 'Business ...
2       ['Software Development', 'Machine Learning', '...
3       ['accounts payables', 'accounts receivables', ...
4       ['Analytical reasoning', 'Compliance testing k...
                              ...                        
9539    ['Mathematical modelling', 'Machine Learning',...
9540    ['Data Analysis', 'Business Analysis', 'Machin...
9541    ['Business Analyst', 'Data Analytics', 'Data C...
9542    ['Machine Learning', 'Natural Language Process...
9543    ['assisted living', 'interpersonal and communi...
Name: skills, Length: 9544, dtype: object
0                                          [['Big Data']]
1       [['Data Analysis', 'Business Analysis', 'Machi...
2       [['Unified Payment Interface', 'Risk Predictio...
3       [['accounts receivables', 'banking', 'G/L Acco...
4       [['collections', 'accounts receivable', 'finan...
                              

#### skills and related_silss_in_job implementation
Implementation Steps

1-Convert skills and related_skils_in_job from string format to Python lists.

2-Flatten related_skils_in_job to remove extra nesting ([['skill1', 'skill2']] → ['skill1', 'skill2']).

3-Merge the cleaned lists into a single column skills.

4-Remove duplicates within each row.

5-Ensure NaN values are handled properly.

In [402]:
import ast

# Function to safely convert stringified lists to actual lists
def safe_list_conversion(val):
    if isinstance(val, str):
        return ast.literal_eval(val)  # Convert string to list
    elif isinstance(val, list):
        return val  # Already a list
    return []  # Return empty list if NaN or other types

# Apply conversion to both columns
resume_data["skills"] = resume_data["skills"].apply(safe_list_conversion)
resume_data["related_skils_in_job"] = resume_data["related_skils_in_job"].apply(safe_list_conversion)
print(resume_data.skills)
print(resume_data.related_skils_in_job)

0       [Big Data, Hadoop, Hive, Python, Mapreduce, Sp...
1       [Data Analysis, Data Analytics, Business Analy...
2       [Software Development, Machine Learning, Deep ...
3       [accounts payables, accounts receivables, Acco...
4       [Analytical reasoning, Compliance testing know...
                              ...                        
9539    [Mathematical modelling, Machine Learning, Pre...
9540    [Data Analysis, Business Analysis, Machine Lea...
9541    [Business Analyst, Data Analytics, Data Cleans...
9542    [Machine Learning, Natural Language Processing...
9543    [assisted living, interpersonal and communicat...
Name: skills, Length: 9544, dtype: object
0                                            [[Big Data]]
1       [[Data Analysis, Business Analysis, Machine Le...
2       [[Unified Payment Interface, Risk Prediction, ...
3       [[accounts receivables, banking, G/L Accounts,...
4       [[collections, accounts receivable, financial ...
                              

In [403]:
# Flatten nested lists in related_skils_in_job
def flatten_list(nested_list):
    flat_list = []
    for sublist in nested_list:
        if isinstance(sublist, list):  # Check if it's a list inside list
            flat_list.extend(sublist)  # Flatten
        else:
            flat_list.append(sublist)  # Single element case
    return flat_list

resume_data["related_skils_in_job"] = resume_data["related_skils_in_job"].apply(flatten_list)
print(resume_data.related_skils_in_job)

0                                              [Big Data]
1       [Data Analysis, Business Analysis, Machine Lea...
2       [Unified Payment Interface, Risk Prediction, B...
3       [accounts receivables, banking, G/L Accounts, ...
4       [collections, accounts receivable, financial r...
                              ...                        
9539    [Data Sampling, Management of Agricultural Pro...
9540                  [EDA, Data Cleaning, Data Modeling]
9541                                   [Machine Learning]
9542               [NLP, Machine Learning, Text Analysis]
9543    [building referral relationships, management, ...
Name: related_skils_in_job, Length: 9544, dtype: object


In [404]:
# Merge both columns
resume_data["skills"] = resume_data.apply(lambda row: list(set(row["skills"] + row["related_skils_in_job"])), axis=1)

# Drop old related_skils_in_job column if no longer needed
resume_data.drop(columns=["related_skils_in_job"], inplace=True)

# Show cleaned data
print(resume_data["skills"][0])

['Hdfs', 'Core Java', 'Hive', 'Talend', 'YARN', 'Microsoft Azure', 'Java', 'Hadoop', 'Data Structures', 'Machine Learning', 'Cloud', 'Amazon Redshift', 'RDBMS', 'Python', 'Informatica', 'Data Science', 'C++', 'Spark', 'Mapreduce', 'DBMS', 'Big Data']


In [405]:
print(resume_data.languages.value_counts())
print(resume_data.proficiency_levels.value_counts())

languages
['English', 'Spanish']                         168
['Spanish']                                     84
['English', 'French']                           28
['Portuguese']                                  28
['English', 'Marathi', 'Hindi', 'Gujarati']     28
['English', 'Urdu', 'German']                   28
['English', 'French', 'Spanish']                28
['English', 'Arabic', 'Swedish']                28
['Arabic', 'English']                           28
['English', 'Chinese Mandarin']                 28
['English', 'Mandarin', 'Cantonese']            28
['English', 'Swahili', 'Kalenjin']              28
['English', 'French', 'Italian', 'Greek']       28
['Spanish', 'English', 'Portuguese']            28
['Japanese', 'English']                         28
['Polish']                                      28
['English', 'Hebrew']                           28
['English', 'Hindi', 'Malayalam']               28
Name: count, dtype: int64
proficiency_levels
['Fluent', 'Fluent', 'Fluen

### languages 

In [406]:
print(resume_data.languages)
print(resume_data.proficiency_levels)

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
9539    NaN
9540    NaN
9541    NaN
9542    NaN
9543    NaN
Name: languages, Length: 9544, dtype: object
0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
9539    NaN
9540    NaN
9541    NaN
9542    NaN
9543    NaN
Name: proficiency_levels, Length: 9544, dtype: object


#### make language and proficiency_levels
Implementation Plan

Convert both columns from string format to lists.

Standardize proficiency levels (e.g., "Fluent - Full Knowledge" → "Fluent").

Pair each language with its corresponding proficiency level (if available).

Create a final merged languages column in the format:

  "English (Fluent), Spanish (Intermediate)"
  
  "Arabic (Native), French (Basic)"


In [407]:
# Function to safely convert stringified lists to actual lists
def safe_list_conversion(val):
    if isinstance(val, str):
        try:
            return ast.literal_eval(val)  # Convert string to list
        except:
            return []  # Handle invalid formats
    elif isinstance(val, list):
        return val  # Already a list
    return []  # Return empty list if NaN or other types

# Convert columns
resume_data["languages"] = resume_data["languages"].apply(safe_list_conversion)
resume_data["proficiency_levels"] = resume_data["proficiency_levels"].apply(safe_list_conversion)
print(resume_data["languages"].value_counts())
print(resume_data["proficiency_levels"].value_counts())

languages
[]                                     8844
[English, Spanish]                      168
[Spanish]                                84
[Arabic, English]                        28
[English, French]                        28
[Portuguese]                             28
[English, Marathi, Hindi, Gujarati]      28
[English, Urdu, German]                  28
[English, French, Spanish]               28
[English, Arabic, Swedish]               28
[English, Mandarin, Cantonese]           28
[English, Chinese Mandarin]              28
[English, Swahili, Kalenjin]             28
[English, French, Italian, Greek]        28
[Spanish, English, Portuguese]           28
[Japanese, English]                      28
[Polish]                                 28
[English, Hebrew]                        28
[English, Hindi, Malayalam]              28
Name: count, dtype: int64
proficiency_levels
[]                                                                                8844
[Fluent, Fluent, Fluen

In [408]:
# Standardize proficiency levels
def clean_proficiency(proficiency_list):
    mapping = {
        "Fluent - Full Knowledge": "Fluent",
        "Fluent; speak, read, and write": "Fluent",
        "Fluent": "Fluent",
        "fluent": "Fluent",
        "Bi-lingual": "Bilingual",
        "Bilingual": "Bilingual",
        "mother tongue": "Native",
        "Native Speaker": "Native",
        "Native": "Native",
        "Professional working proficiency": "Professional",
        "Decreased proficiency due to moderate use": "Limited",
        "Knowledge, but taking classes to become more fluent": "Basic",
        "Fluent reading/writing": "Intermediate",
        "GCSE level": "Basic",
        "Good in both written and spoken": "Intermediate",
        "N/A": None,
        None: None
    }
    return [mapping.get(level, level) for level in proficiency_list if mapping.get(level, level) is not None]

resume_data["proficiency_levels"] = resume_data["proficiency_levels"].apply(clean_proficiency)
print(resume_data["proficiency_levels"].value_counts())

proficiency_levels
[]                                      8956
[Fluent]                                  84
[Fluent, Fluent, Fluent]                  84
[Bilingual, Bilingual]                    84
[Native, Native]                          28
[Fluent, Limited]                         28
[Fluent, Native]                          28
[Proficient]                              28
[Native, Professional]                    28
[Bilingual, Bilingual, Intermediate]      28
[Fluent, Fluent, Fluent, Fluent]          28
[Native, Intermediate]                    28
[Native]                                  28
[Advanced, Fluent]                        28
[Fluent, Fluent, Basic]                   28
[Fluent, Basic]                           28
Name: count, dtype: int64


In [409]:
# Merge languages with their proficiency levels
def merge_languages_proficiency(langs, profs):
    merged = []
    for i in range(len(langs)):
        level = profs[i] if i < len(profs) else "Unknown"  # Assign "Unknown" if missing
        merged.append(f"{langs[i]} ({level})")
    return ", ".join(merged)

resume_data["languages"] = resume_data.apply(lambda row: merge_languages_proficiency(row["languages"], row["proficiency_levels"]), axis=1)

# Drop old proficiency_levels column if no longer needed
resume_data.drop(columns=["proficiency_levels"], inplace=True)

# Show cleaned data
print(resume_data["languages"].head())

0    
1    
2    
3    
4    
Name: languages, dtype: object


In [410]:
resume_data.languages.value_counts()

languages
                                                                             8844
Spanish (Fluent)                                                               56
English (Bilingual), Spanish (Bilingual)                                       56
English (Fluent), Spanish (Limited)                                            28
English (Fluent), Arabic (Fluent), Swedish (Fluent)                            28
English (Unknown), Hindi (Unknown), Malayalam (Unknown)                        28
English (Fluent), Spanish (Basic)                                              28
English (Unknown), French (Unknown)                                            28
Portuguese (Fluent)                                                            28
English (Unknown), Marathi (Unknown), Hindi (Unknown), Gujarati (Unknown)      28
English (Fluent), Urdu (Fluent), German (Basic)                                28
English (Advanced), Spanish (Fluent)                                           28
Englis

In [411]:
resume_data.shape

(9544, 18)

## location

In [412]:
resume_data.address.value_counts()

address
Sunnyvale, CA                                                                                                     28
Trenton, NJ                                                                                                       28
Malvern, PA                                                                                                       28
Department of Computer Science, University of British Columbia, 2366 Main Mall, Vancouver, BC, V6T 1Z4, Canada    28
Hoboken, NJ                                                                                                       28
Hanover, NH                                                                                                       28
Denver, CO                                                                                                        28
Pittsburgh, PA                                                                                                    28
Chicago, IL                                             

Extract City and State/Country:

If the address contains City, State (e.g., Chicago, IL) → Keep it as is.

If the address contains a full address (e.g., 2383 Suwanee Pointe Drive, Lawrenceville, GA 30043) → Extract City and State.

If the address contains a university department address (e.g., Department of Computer Science, University of British Columbia, Vancouver, BC, Canada) → Extract City and Country.

If the address contains a P.O. Box or generic placeholder (e.g., P.O BOX 2200 ELDORET) → Extract the city (Eldoret in this case).


If the address is ambiguous (e.g., Address, City, Province, Country) → Drop or infer from another column if possible.


Normalize formats:

Ensure consistency between state abbreviations and full names.

Convert country names to ISO country codes if needed.

Handle missing or generic values:

If an address is missing or too vague, try filling it from another column (e.g., education history or company name).

In [413]:
import re
import pandas as pd

def extract_location(address):
    """Extracts city, state, or country from a given address."""
    if pd.isna(address) or 'Address' in address or 'City' in address:
        return None  # Skip generic placeholders

    # Regex for extracting City, State format (e.g., 'Chicago, IL' or 'Vancouver, BC')
    city_state_pattern = r'([A-Za-z\s]+),\s?([A-Z]{2})\b'
    match = re.search(city_state_pattern, address)
    if match:
        return f"{match.group(1).strip()}, {match.group(2).strip()}"

    # Extract full country name if present at the end
    country_pattern = r'([A-Za-z\s]+),\s?([A-Za-z\s]+),\s?([A-Za-z]+)$'
    country_match = re.search(country_pattern, address)
    if country_match:
        return f"{country_match.group(2).strip()}, {country_match.group(3).strip()}"

    # Extract just city if no state/country is found
    city_only_pattern = r'([A-Za-z\s]+)\s?\d*$'
    city_match = re.search(city_only_pattern, address)
    if city_match:
        return city_match.group(1).strip()

    return None  # Default case if no match

# Apply cleaning function to the dataset
resume_data['location'] = resume_data['address'].apply(extract_location)
resume_data.drop(columns=['address'])

# Display the first few cleaned locations
print(resume_data[['location']].value_counts())

location                                                        
Vancouver, BC                                                       56
Boca Raton, FL                                                      28
Cambridge, MA                                                       28
Virginia Beach, VA                                                  28
Trenton, NJ                                                         28
Sunnyvale, CA                                                       28
Portland, OR                                                        28
Pittsburgh, PA                                                      28
P                                                                   28
Malvern, PA                                                         28
Lawrenceville, GA                                                   28
Las Vegas, NV                                                       28
Hoboken, NJ                                                         28
Hanover, NH 

### QUalification column from --> (carrer_objective , education_instiuation_name , degree_name, major_field_of_studies)

In [414]:
print(resume_data.career_objective.value_counts())

career_objective
Big data analytics working and database warehouse manager with robust experience in handling all kinds of data. I have also used multiple cloud infrastructure services and am well acquainted with them. Currently in search of role that offers more of development.                                                               28
To utilize my customer relations and management background to obtain a position with a company that places importance on efficiency and success. Employ natural leadership and interpersonal skills to build relationships throughout the organization.                                                                              28
Fresher python developer interested in data analytics and machine learning.                                                                                                                                                                                                                                                          28

In [415]:
print(resume_data.educational_institution_name.value_counts())

educational_institution_name
['Illinois Institute of Technology', 'National Institute of Technology']                                                                                            112
['Nagpur University']                                                                                                                                                84
['National Institute of Technology (NIT), Hamirpur']                                                                                                                 56
['New York University']                                                                                                                                              56
['VIT-AP University']                                                                                                                                                56
                                                                                                                                   

In [416]:
print(resume_data.degree_names.value_counts())

degree_names
['B.Tech']                                                                      1764
['Bachelor of Science']                                                          980
['B.Tech', 'M.Tech']                                                             308
['B.S.']                                                                         168
['Master of Science', 'Bachelor of Science']                                     168
                                                                                ... 
['Course Revisions for the: National Electrical Code', 'Associate Degree']        28
['Certified Professional Coach', 'Bachelor of Science']                           20
['Masters of Business Administration', 'Bachelors of Science', 'N/A']             12
['ACCA II Part Qualified', 'Diploma', 'Grade 12 (School Certificate)', None]       6
['Master of Science Degree', 'Bachelor of Science']                                4
Name: count, Length: 180, dtype: int64


In [417]:
print(resume_data.major_field_of_studies.value_counts())

major_field_of_studies
['N/A']                                                                 616
['Computer Science']                                                    560
['Computers']                                                           448
['Accounting']                                                          280
['ECE']                                                                 168
                                                                       ... 
['Coaching', 'Business Administration, Marketing']                       20
['Business Strategy', 'Business Management', 'Biology/Pre-Medicine']     12
['Administrative Assistance']                                            10
[None, 'Accountancy', None, None]                                         6
['Psychology', 'Integrated Marketing Communications']                     4
Name: count, Length: 210, dtype: int64


#### on carrer_objective column 

In [418]:
def clean_text(text):
    if pd.isna(text):  # Handle NaN values
        return ""
    
    text = text.lower().strip()  # Convert to lowercase and strip spaces
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    text = re.sub(r'[^\w\s]', '', text)  # Remove special characters except spaces
    return text

# Apply cleaning function
resume_data['career_objective'] = resume_data['career_objective'].apply(clean_text)
print(resume_data['career_objective'].value_counts())


career_objective
                                                                                                                                                                                                                                                                                                                                  4804
big data analytics working and database warehouse manager with robust experience in handling all kinds of data i have also used multiple cloud infrastructure services and am well acquainted with them currently in search of role that offers more of development                                                                 28
to utilize my customer relations and management background to obtain a position with a company that places importance on efficiency and success employ natural leadership and interpersonal skills to build relationships throughout the organization                                                                               28
fr

### on educational_institution_name

In [419]:
import ast

def clean_educational_institution_name(name):
    if pd.isna(name):  # التحقق مما إذا كانت القيمة فارغة
        return ""  # استبدال القيم الفارغة بسلسلة فارغة

    if isinstance(name, str):
        try:
            name_list = ast.literal_eval(name)  # تحويل النص إلى قائمة
            if isinstance(name_list, list):
                name_list = [str(item).strip() for item in name_list if item]  # إزالة الفراغات والتأكد من أن القيم ليست فارغة
                cleaned_name = ', '.join(name_list)
            else:
                cleaned_name = name.strip()
        except (ValueError, SyntaxError):
            cleaned_name = name.strip()  # إبقاء النص كما هو إذا لم يكن بصيغة قائمة
    else:
        cleaned_name = str(name).strip()  # تحويل أي قيمة أخرى إلى نص وإزالة الفراغات

    return cleaned_name

# Apply the function to the column
resume_data['educational_institution_name'] = resume_data['educational_institution_name'].apply(clean_educational_institution_name)

print(resume_data.educational_institution_name.value_counts())

educational_institution_name
Illinois Institute of Technology, National Institute of Technology                                                                                        112
                                                                                                                                                           84
Nagpur University                                                                                                                                          84
New York University                                                                                                                                        56
National Institute of Technology (NIT), Hamirpur                                                                                                           56
                                                                                                                                                         ... 
Institute of Profession

### on degree_names

In [420]:
print(resume_data['degree_names'].isna().sum())  # عدد القيم الفارغة


84


In [421]:
print(resume_data['degree_names'].value_counts())


degree_names
['B.Tech']                                                                      1764
['Bachelor of Science']                                                          980
['B.Tech', 'M.Tech']                                                             308
['B.S.']                                                                         168
['Master of Science', 'Bachelor of Science']                                     168
                                                                                ... 
['Course Revisions for the: National Electrical Code', 'Associate Degree']        28
['Certified Professional Coach', 'Bachelor of Science']                           20
['Masters of Business Administration', 'Bachelors of Science', 'N/A']             12
['ACCA II Part Qualified', 'Diploma', 'Grade 12 (School Certificate)', None]       6
['Master of Science Degree', 'Bachelor of Science']                                4
Name: count, Length: 180, dtype: int64


In [422]:
print(resume_data['degree_names'].apply(type).value_counts())

degree_names
<class 'str'>      9460
<class 'float'>      84
Name: count, dtype: int64


 مشاكل موجودة في البيانات:
وجود قيم فارغة (NaN أو None)

هناك 84 قيمة مفقودة (NaN)، ويمكن استبدالها بـ "Unknown" أو "N/A" حسب الحاجة.
القيم مكتوبة كقوائم (list) بدلاً من نصوص (str)

العديد من القيم مثل ['B.Tech', 'M.Tech'] هي قوائم داخل الخلايا، مما يعيق عمليات البحث والتحليل.
تحتاج هذه القوائم إلى التحويل إلى نص عادي مثل "B.Tech, M.Tech".
بعض القيم تحتوي على None داخل القوائم

مثال: ['ACCA II Part Qualified', 'Diploma', 'Grade 12 (School Certificate)', None]
يجب إزالة None من القوائم عند تحويلها إلى نصوص.

In [423]:
# تنظيف القيم في عمود degree_names
def clean_degree_names(value):
    if pd.isna(value):
        return "Unknown"  # استبدال القيم الفارغة
    
    if isinstance(value, str):
        try:
            # محاولة تحويل النص إلى قائمة إذا كان بتنسيق list
            value_list = ast.literal_eval(value)  
            if isinstance(value_list, list):
                # إزالة None وتحويل القائمة إلى نص مفصول بفواصل
                return ", ".join([item for item in value_list if item is not None])
        except (SyntaxError, ValueError):
            pass  # إذا فشل التحويل، نترك القيمة كما هي

    return value  # إعادة القيمة الأصلية إذا لم تكن بحاجة إلى تعديل

# تطبيق الدالة على العمود
resume_data["degree_names"] = resume_data["degree_names"].apply(clean_degree_names)

# التحقق من النتيجة بعد التنظيف
print(resume_data["degree_names"].value_counts())

degree_names
B.Tech                                                                  1764
Bachelor of Science                                                      980
B.Tech, M.Tech                                                           308
Master of Science, Bachelor of Science                                   168
B.S.                                                                     168
                                                                        ... 
Course Revisions for the: National Electrical Code, Associate Degree      28
Certified Professional Coach, Bachelor of Science                         20
Masters of Business Administration, Bachelors of Science, N/A             12
ACCA II Part Qualified, Diploma, Grade 12 (School Certificate)             6
Master of Science Degree, Bachelor of Science                              4
Name: count, Length: 178, dtype: int64


In [424]:
# التحقق من عدد القيم الفارغة
print("عدد القيم الفارغة:", resume_data["major_field_of_studies"].isna().sum())

# التحقق من أنواع البيانات في العمود
print(resume_data["major_field_of_studies"].apply(type).value_counts())

# عرض أكثر القيم تكرارًا لرؤية شكل البيانات
print(resume_data["major_field_of_studies"].value_counts().head(20))


عدد القيم الفارغة: 84
major_field_of_studies
<class 'str'>      9460
<class 'float'>      84
Name: count, dtype: int64
major_field_of_studies
['N/A']                                                              616
['Computer Science']                                                 560
['Computers']                                                        448
['Accounting']                                                       280
['ECE']                                                              168
['Information Technology']                                           168
['Electrical']                                                       140
['Mechanical Engineering']                                           140
['CSE']                                                              140
[None]                                                               112
['Computer Science', 'Electronics and Communication Engineering']    112
['Accounting', 'Accounting']                           

Name: count, dtype: int64
ChatGPT said:
🔍 تحليل عمود major_field_of_studies
نعم، هذا العمود يحتاج إلى تنظيف بسبب المشاكل التالية:

وجود قيم فارغة (NaN / None)

يوجد 84 قيمة فارغة (NaN) و 112 قيمة تحتوي على None داخل القوائم.
يجب استبدال القيم الفارغة بـ "Unknown" أو "N/A" لضمان تناسق البيانات.
وجود قوائم بدلاً من نصوص (['Computer Science', 'Mathematics'])

يجب تحويل القوائم إلى نصوص مفصولة بفواصل (", ").
وجود تكرارات داخل القائمة (['Computer Science', 'Computer Science'])

يجب إزالة التكرارات داخل القائمة.
وجود قيم غير واضحة (['N/A'], ['IT'], ['CSE'])

يمكن توحيد بعض القيم المتشابهة (['IT'] → "Information Technology", ['CSE'] → "Computer Science").

In [425]:
def clean_major_field(field):
    if pd.isna(field):  # استبدال القيم الفارغة
        return "Unknown"
    
    try:
        field_list = ast.literal_eval(field) if isinstance(field, str) else field
        if isinstance(field_list, list):
            field_list = [item for item in field_list if item and item != "None"]  # إزالة None
            field_list = list(set(field_list))  # إزالة التكرارات
            return ", ".join(field_list) if field_list else "Unknown"
    except:
        return field  # في حالة حدوث خطأ، نعيد القيمة الأصلية
    
    return field  # إذا لم يكن هناك حاجة للتعديل

# تطبيق التنظيف على العمود
resume_data["major_field_of_studies"] = resume_data["major_field_of_studies"].apply(clean_major_field)

# عرض النتائج بعد التنظيف
print(resume_data["major_field_of_studies"].value_counts().head(20))

major_field_of_studies
Computer Science                                               728
N/A                                                            728
Computers                                                      504
Accounting                                                     392
Unknown                                                        308
Mechanical Engineering                                         196
ECE                                                            168
Electrical                                                     168
Information Technology                                         168
CSE                                                            140
Electronics and Communication Engineering, Computer Science    112
Civil                                                          112
IT                                                             112
Electronics/Telecommunication                                  112
Economics                              

In [426]:
# دمج الأعمدة الأربعة في عمود واحد باسم qualification
resume_data["qualification"] = resume_data[["career_objective", "educational_institution_name", "degree_names", "major_field_of_studies"]] \
    .astype(str) \
    .agg(lambda x: ", ".join(x.dropna().unique()), axis=1)

# حذف الأعمدة الأصلية بعد الدمج
resume_data.drop(columns=["career_objective", "educational_institution_name", "degree_names", "major_field_of_studies"], inplace=True)

# عرض أول 5 صفوف للتأكد من الدمج الصحيح
print(resume_data.head())

  address                                             skills  \
0     NaN  [Hdfs, Core Java, Hive, Talend, YARN, Microsof...   
1     NaN  [R, SAS, Data Visualization, Business Analytic...   
2     NaN  [HTML, Hive, PySpark, Application Support, Uni...   
3     NaN  [commercial auto underwriting, mortgage loan, ...   
4     NaN  [administrative, audit procedures, general led...   

  educational_results    result_types  \
0             ['N/A']          [None]   
1      ['N/A', 'N/A']  ['N/A', 'N/A']   
2             ['N/A']         ['N/A']   
3              [None]          [None]   
4            ['3.84']          [None]   

                                         start_dates  \
0                                       ['Nov 2019']   
1                                       ['Sep 2019']   
2                                      ['June 2018']   
3  ['January 2011', 'January 2008', 'January 2006...   
4  ['January 2016', 'January 2016', 'January 2012...   

                               

In [427]:
resume_data.head(2)

address                                             skills  \
0     NaN  [Hdfs, Core Java, Hive, Talend, YARN, Microsof...   
1     NaN  [R, SAS, Data Visualization, Business Analytic...   

  educational_results    result_types   start_dates      end_dates  \
0             ['N/A']          [None]  ['Nov 2019']  ['Till Date']   
1      ['N/A', 'N/A']  ['N/A', 'N/A']  ['Sep 2019']  ['Till Date']   

              positions                                   responsibilities  \
0  ['Big Data Analyst']  Technical Support\nTroubleshooting\nCollaborat...   
1  ['Business Analyst']  Machine Learning Leadership\nCross-Functional ...   

  extra_curricular_activity_types extra_curricular_organization_names  \
0                             NaN                                 NaN   
1                             NaN                                 NaN   

  extra_curricular_organization_links role_positions languages  \
0                                 NaN            NaN             
1                                 NaN            NaN             

  certification_providers location  \
0                     NaN     None   
1                     NaN     None   

                                       qualification  
0  big data analytics working and database wareho...  
1  fresher looking to join as a data analyst and ...

## convert end_dates and start_dates to experiecne year column

In [428]:
for i in range(300):
  print(resume_data.start_dates[i])
  print(resume_data.end_dates[i])


['Nov 2019']
['Till Date']
['Sep 2019']
['Till Date']
['June 2018']
['Till Date']
['January 2011', 'January 2008', 'January 2006', 'January 2004', 'January 2001', 'N/A']
['November 2015', 'January 2010', 'January 2008', 'January 2006', 'January 2004', None]
['January 2016', 'January 2016', 'January 2012', 'January 2009', 'January 2006']
['Current', 'January 2016', 'January 2015', 'January 2011', 'January 2009']
['August 2006', 'August 2013', 'July 2014']
['January 2013', 'September 2014', 'Current']
['Jan 2019']
['Sep 2019']
['January 2013', 'January 2011', 'January 2005']
['Current', 'January 2012', 'January 2011']
['Jun 2019']
['till date']
['Jan 2019']
['Till Date']
['January 1995', 'January 2003', 'January 1993']
['Current', 'January 2004', 'January 1995']
['APR 2020']
['TILL DATE']
['Jul 2020', 'Dec 2019']
['till date', 'May 2020']
['July 2021', 'May 2021']
['Aug 2021', 'June 2021']
['May 2021', 'January 2021']
['current', 'May 2021']
['Jun 2020']
['till date']
['01/2014', '12/201

In [429]:
resume_data.start_dates.value_counts()

start_dates
['Nov 2019']                                                                                           280
['Jan 2019']                                                                                           280
['Apr 2019']                                                                                           280
['May 2019']                                                                                           252
['Jan 2020']                                                                                           196
                                                                                                      ... 
['03/2014', '06/2003', '01/2002', '01/2001', '01/1995', '01/1991']                                      20
['06/2014', '06/2014', '07/2014', '08/2011', '06/2009', '12/2007', '01/2004', '03/1999', '04/1996']     12
['Jul 1996', 'Nov 1995', 'Oct 1993', 'Jan 1990']                                                        10
['09/2005', '01/2004', '0

In [430]:
resume_data.end_dates.value_counts()

end_dates
['Till Date']                                                                                          700
['Ongoing']                                                                                            588
['till date']                                                                                          336
['Oct 2020']                                                                                           280
['Dec 2020']                                                                                           168
                                                                                                      ... 
['10/2019', '11/2013', '01/2003', '01/2002', '01/2000', '01/1995']                                      20
['Current', 'Current', 'Current', '06/2014', '08/2011', '08/2009', '11/2007', '02/2004', '03/1999']     12
['Jul 2016', 'Jul 1996', 'Nov 1995', 'Oct 1992']                                                        10
['Current', '05/2005', '12/

In [431]:
resume_data.shape

(9544, 16)

In [432]:
resume_data.columns

Index(['address', 'skills', 'educational_results', 'result_types',
       'start_dates', 'end_dates', 'positions', 'responsibilities',
       'extra_curricular_activity_types',
       'extra_curricular_organization_names',
       'extra_curricular_organization_links', 'role_positions', 'languages',
       'certification_providers', 'location', 'qualification'],
      dtype='object')

Convert Strings to Lists Properly:

Some values are single dates ('Nov 2019'), while others are lists ("['03/2014', '06/2003', '01/2002']").
We need to extract dates correctly from string representations of lists.
Standardize Date Formats:

Some dates are in "MMM YYYY" format (e.g., 'Nov 2019'), while others are in "MM/YYYY" format (e.g., '03/2014').
We'll parse both formats correctly.
Extract Minimum Date for Each Row:

If a row contains multiple dates, we'll find the earliest (oldest) date.
Handle Missing or Invalid Data:

Convert any NaN or empty lists to NaT (Not a Time).
Handle errors gracefully during conversion.

In [433]:
import pandas as pd
import ast  # To safely convert string representations of lists

def extract_min_date(date_list):
    if pd.isna(date_list) or date_list == "[]":
        return pd.NaT  # Return NaT if empty or NaN
    
    if isinstance(date_list, str):  
        try:
            date_list = ast.literal_eval(date_list)  # Convert string to list
        except (ValueError, SyntaxError):
            date_list = [date_list]  # If it's a single string, treat it as a list
    
    if not isinstance(date_list, list):
        date_list = [date_list]  # Ensure it's a list
    
    parsed_dates = []
    for date in date_list:
        date = str(date).strip()
        
        # Try parsing different date formats
        try:
            if "/" in date:  # Format: MM/YYYY
                parsed_dates.append(pd.to_datetime(date, format="%m/%Y", errors="coerce"))
            else:  # Format: MMM YYYY (e.g., 'Nov 2019')
                parsed_dates.append(pd.to_datetime(date, format="%b %Y", errors="coerce"))
        except Exception:
            continue  # Skip invalid dates
    
    # Return the minimum valid date
    return min(parsed_dates) if parsed_dates else pd.NaT

# Apply the function to the column
resume_data["start_dates"] = resume_data["start_dates"].apply(extract_min_date)

# Display cleaned data
resume_data.start_dates.value_counts()

start_dates
2019-01-01    336
2019-11-01    280
2019-04-01    280
2019-05-01    252
2019-12-01    224
             ... 
2011-06-01     28
2021-01-01     28
2008-11-01     28
1991-01-01     20
1996-04-01     12
Name: count, Length: 116, dtype: int64

Steps for Data Cleaning
Convert Strings to Lists Properly:

Some values are single dates ('Oct 2020'), while others are lists ("['10/2019', '11/2013', '01/2003']").
We need to extract dates correctly from string representations of lists.
Standardize Date Formats:

Some dates are in "MMM YYYY" format (e.g., 'Oct 2020'), while others are in "MM/YYYY" format (e.g., '10/2019').
We'll parse both formats correctly.
Handle "Current" Values:

Values like 'Till Date', 'Ongoing', 'Current', and variations like 'till date' should be converted to NaT (Not a Time) or today's date.
Extract Maximum Date for Each Row:

If a row contains multiple dates, we'll find the latest (most recent) date.
Handle Missing or Invalid Data:

Convert any NaN or empty lists to NaT.
Handle errors gracefully during conversion.

In [434]:
import pandas as pd
import ast  # To safely convert string representations of lists
from datetime import datetime

def extract_max_date(date_list):
    if pd.isna(date_list) or date_list == "[]":
        return pd.NaT  # Return NaT if empty or NaN
    
    if isinstance(date_list, str):  
        try:
            date_list = ast.literal_eval(date_list)  # Convert string to list
        except (ValueError, SyntaxError):
            date_list = [date_list]  # If it's a single string, treat it as a list
    
    if not isinstance(date_list, list):
        date_list = [date_list]  # Ensure it's a list
    
    parsed_dates = []
    for date in date_list:
        date = str(date).strip()

        # Handle 'Current' values
        if date.lower() in ['till date', 'ongoing', 'current']:
            parsed_dates.append(pd.Timestamp.today())  # Use today's date

        else:
            # Try parsing different date formats
            try:
                if "/" in date:  # Format: MM/YYYY
                    parsed_dates.append(pd.to_datetime(date, format="%m/%Y", errors="coerce"))
                else:  # Format: MMM YYYY (e.g., 'Oct 2020')
                    parsed_dates.append(pd.to_datetime(date, format="%b %Y", errors="coerce"))
            except Exception:
                continue  # Skip invalid dates
    
    # Return the maximum valid date
    return max(parsed_dates) if parsed_dates else pd.NaT

# Apply the function to the column
resume_data["end_dates"] = resume_data["end_dates"].apply(extract_max_date)

# Display cleaned data
resume_data.end_dates.value_counts()

end_dates
2020-10-01 00:00:00.000000    308
2020-12-01 00:00:00.000000    196
2016-01-01 00:00:00.000000    140
2020-04-01 00:00:00.000000    112
2020-09-01 00:00:00.000000     84
                             ... 
2025-02-09 01:46:56.261954      1
2025-02-09 01:46:57.236204      1
2025-02-09 01:46:57.237204      1
2025-02-09 01:46:57.238205      1
2025-02-09 01:46:55.438080      1
Name: count, Length: 2212, dtype: int64

In [435]:
# Ensure start_dates and end_dates are in datetime format
resume_data["start_dates"] = pd.to_datetime(resume_data["start_dates"])
resume_data["end_dates"] = pd.to_datetime(resume_data["end_dates"])

# Calculate the difference in years
resume_data["year_of_experience"] = (resume_data["end_dates"] - resume_data["start_dates"]).dt.days / 365.25

# Display the updated dataframe
print(resume_data.year_of_experience.value_counts())


year_of_experience
5.275838     168
6.108145     168
5.108830     140
0.585900     140
5.779603     140
            ... 
23.832991     28
14.611910     28
28.747433     20
28.859685     12
26.496920     10
Name: count, Length: 143, dtype: int64


In [436]:
# Drop the original date columns
resume_data.drop(columns=["start_dates", "end_dates"], inplace=True)

In [437]:
resume_data.head()

address                                             skills  \
0     NaN  [Hdfs, Core Java, Hive, Talend, YARN, Microsof...   
1     NaN  [R, SAS, Data Visualization, Business Analytic...   
2     NaN  [HTML, Hive, PySpark, Application Support, Uni...   
3     NaN  [commercial auto underwriting, mortgage loan, ...   
4     NaN  [administrative, audit procedures, general led...   

  educational_results    result_types  \
0             ['N/A']          [None]   
1      ['N/A', 'N/A']  ['N/A', 'N/A']   
2             ['N/A']         ['N/A']   
3              [None]          [None]   
4            ['3.84']          [None]   

                                           positions  \
0                               ['Big Data Analyst']   
1                               ['Business Analyst']   
2  ['Software Developer (Machine Learning Enginee...   
3  ['Accountant', 'Accounts Receivable Clerk', 'M...   
4  ['Staff Accountant', 'Senior Accountant', 'Tax...   

                                    responsibilities  \
0  Technical Support\nTroubleshooting\nCollaborat...   
1  Machine Learning Leadership\nCross-Functional ...   
2  Trade Marketing Executive\nBrand Visibility, S...   
3  Apparel Sourcing\nQuality Garment Sourcing\nRe...   
4  iOS Lifecycle\nRequirement Analysis\nNative Fr...   

                     extra_curricular_activity_types  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  ['Professional Organization', 'Honor Society',...   

                 extra_curricular_organization_names  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  ['Ohio Society of CPAs', 'Beta Alpha Psi', 'Go...   

  extra_curricular_organization_links            role_positions languages  \
0                                 NaN                       NaN             
1                                 NaN                       NaN             
2                                 NaN                       NaN             
3                                 NaN                       NaN             
4            [None, None, None, None]  [None, None, None, None]             

  certification_providers location  \
0                     NaN     None   
1                     NaN     None   
2                     NaN     None   
3                     NaN     None   
4  ['Ohio Notary Public']     None   

                                       qualification  year_of_experience  
0  big data analytics working and database wareho...            5.275838  
1  fresher looking to join as a data analyst and ...            5.442847  
2  , Birla Institute of Technology (BIT), Ranchi,...                 NaN  
3  to obtain a position in a fastpaced business o...                 NaN  
4  professional accountant with an outstanding wo...                 NaN

## on work history column form --> positions , role_positions, responsibilities

In [438]:
resume_data.positions.value_counts()

positions
['Intern']                                                                                                                                                                                                                                                                             392
['Intern Trainee']                                                                                                                                                                                                                                                                     140
['Data Analyst']                                                                                                                                                                                                                                                                       112
['Software Engineering Manager', 'Software Engineering Manager', 'Principal Software Engineer', 'Software Engineer']                         

In [439]:
resume_data.role_positions.value_counts()

role_positions
['N/A']                                                              532
[None]                                                               168
['Participant']                                                      140
['RUNNER-UP, SOLINOVATE']                                             56
[None, None]                                                          56
                                                                    ... 
['Team lead of Enactus', 'Coordinator', 'Senior member']              28
['Best College Elocutioner']                                          28
['PyTorch Developer']                                                 28
['Python statistics with Pandas and Matplotlib']                      28
['Volunteer', 'Volunteer', 'Volunteer', 'Volunteer', 'Volunteer']      4
Name: count, Length: 91, dtype: int64

In [440]:
resume_data.responsibilities.value_counts()

responsibilities
Supervision\nMonitoring\nConstruction\nEstimation\nPlanning\nMaterial Management\nProject Coordination\nQuality Assurance\nCost Control\nInventory\nOperations\nSafety\nError Escalation\nMiscellaneous Tasks\n\n\n\n\n                                                                                                                                                                                                                                                                                                                                                                                                           342
Project Design\nData Analysis\nACCORD/Alliance Knowledge\nBNBC Standards\nCost Estimation\nFeasibility Studies\nDocumentation\nCompliance\nSite Monitoring\nPolicy Enforcement\nLegal Compliance\nData Management\nTeam Collaboration                                                                                                                                            

In [441]:
import pandas as pd

def clean_responsibilities(text):
    if pd.isna(text):  # Handle missing values
        return "Unknown"
    text = text.strip()  # Remove leading/trailing whitespace
    text = text.replace("\n", ", ")  # Convert newlines to commas
    text = ', '.join(set(text.split(", ")))  # Remove duplicate words
    text = text.title()  # Standardize capitalization
    return text

# Apply cleaning function
resume_data["responsibilities"] = resume_data["responsibilities"].apply(clean_responsibilities)

# Display cleaned data
print(resume_data["responsibilities"].head(10))


0    Troubleshooting, System Monitoring, Technical ...
1    Algorithm Analysis, Research Reporting, Protot...
2    Product Distribution, Brand Head, Sales Target...
3    Buyer/Vendor Communication, Apparel Sourcing, ...
4    Task Management, Performance Optimization, Tec...
5    Analytical Skills, Machine Learning Design, Ai...
6    Task Management, Performance Optimization, Tec...
7    Task Management, Performance Optimization, Tec...
8    Troubleshooting, Log Maintenance, Report Prepa...
9    Buyer/Vendor Communication, Apparel Sourcing, ...
Name: responsibilities, dtype: object


In [442]:
# Function to safely convert string lists to actual lists
def safe_eval(val):
    if isinstance(val, str):
        try:
            return ast.literal_eval(val)  # Convert string to list
        except (ValueError, SyntaxError):
            return []
    return val if isinstance(val, list) else []

# Apply safe conversion
resume_data['role_positions'] = resume_data['role_positions'].apply(safe_eval)
resume_data['positions'] = resume_data['positions'].apply(safe_eval)

# Function to clean and merge work history
def merge_work_history(row):
    roles = [r for r in row['role_positions'] if r and r != 'N/A' and r != 'None']
    positions = [p for p in row['positions'] if p and p != 'N/A' and p != 'None']
    responsibilities = row['responsibilities'] if isinstance(row['responsibilities'], str) else ""

    work_history = []
    if roles:
        work_history.append(f"Roles: {', '.join(roles)}")
    if positions:
        work_history.append(f"Positions: {', '.join(positions)}")
    if responsibilities:
        work_history.append(f"Responsibilities: {responsibilities}")

    return ' | '.join(work_history) if work_history else 'No work history available'

# Apply cleaning and merging
resume_data['work_history'] = resume_data.apply(merge_work_history, axis=1)

resume_data.work_history.value_counts()


work_history
Positions: Intern | Responsibilities: Sql Server Clustering & High Availability, Index Rebuilding & Performance Tuning, Etl Process Implementation, High Availability Group Management, Sql Server Replication, Bi Solutions Development, Database Maintenance, Backup & Restore Management, Database Monitoring & Troubleshooting, Sql Query Optimization, Data Integrity & Security, Database Design & Development    5
Positions: Intern | Responsibilities: Stakeholder Collaboration, Data Platform Design, Continuous Learning, Data Quality Monitoring, Etl Processes, Data Integration & Transformation, Sql Query Optimization, Data Modeling, Data Pipeline Development                                                                                                                                                                 5
Positions: Intern | Responsibilities: Troubleshooting, Log Maintenance, Report Preparation, Machinery Maintenance                                                      

In [443]:
resume_data.head(1)

address                                             skills  \
0     NaN  [Hdfs, Core Java, Hive, Talend, YARN, Microsof...   

  educational_results result_types           positions  \
0             ['N/A']       [None]  [Big Data Analyst]   

                                    responsibilities  \
0  Troubleshooting, System Monitoring, Technical ...   

  extra_curricular_activity_types extra_curricular_organization_names  \
0                             NaN                                 NaN   

  extra_curricular_organization_links role_positions languages  \
0                                 NaN             []             

  certification_providers location  \
0                     NaN     None   

                                       qualification  year_of_experience  \
0  big data analytics working and database wareho...            5.275838   

                                        work_history  
0  Positions: Big Data Analyst | Responsibilities...

In [444]:
resume_data.columns

Index(['address', 'skills', 'educational_results', 'result_types', 'positions',
       'responsibilities', 'extra_curricular_activity_types',
       'extra_curricular_organization_names',
       'extra_curricular_organization_links', 'role_positions', 'languages',
       'certification_providers', 'location', 'qualification',
       'year_of_experience', 'work_history'],
      dtype='object')

In [449]:
resume_data.location.value_counts()

location
Vancouver, BC                                                       56
Sunnyvale, CA                                                       28
ELDORET                                                             28
Hoboken, NJ                                                         28
Hanover, NH                                                         28
Denver, CO                                                          28
Pittsburgh, PA                                                      28
Chicago, IL                                                         28
Boca Raton, FL                                                      28
Virginia Beach, VA                                                  28
Lawrenceville, GA                                                   28
Las Vegas, NV                                                       28
Westlake, TX                                                        28
Trenton, NJ                                                         

In [450]:
resume_data = resume_data.drop(columns=['address', 'educational_results', 'result_types', 'positions',
       'responsibilities', 'extra_curricular_activity_types',
       'extra_curricular_organization_names',
       'extra_curricular_organization_links', 'role_positions','certification_providers'])
resume_data.columns

Index(['skills', 'languages', 'location', 'qualification',
       'year_of_experience', 'work_history'],
      dtype='object')

In [451]:
resume_data.head(4)

skills languages location  \
0  [Hdfs, Core Java, Hive, Talend, YARN, Microsof...               None   
1  [R, SAS, Data Visualization, Business Analytic...               None   
2  [HTML, Hive, PySpark, Application Support, Uni...               None   
3  [commercial auto underwriting, mortgage loan, ...               None   

                                       qualification  year_of_experience  \
0  big data analytics working and database wareho...            5.275838   
1  fresher looking to join as a data analyst and ...            5.442847   
2  , Birla Institute of Technology (BIT), Ranchi,...                 NaN   
3  to obtain a position in a fastpaced business o...                 NaN   

                                        work_history  
0  Positions: Big Data Analyst | Responsibilities...  
1  Positions: Business Analyst | Responsibilities...  
2  Positions: Software Developer (Machine Learnin...  
3  Positions: Accountant, Accounts Receivable Cle...

In [452]:
resume_data.shape

(9544, 6)

In [453]:
resume_data.to_csv("secondResumes.csv")

In [454]:
finalResumeSecond = pd.read_csv("secondResumes.csv")

In [455]:
finalResumeSecond.shape

(9544, 7)

In [456]:
finalResumeSecond.dropna()

Empty DataFrame
Columns: [Unnamed: 0, skills, languages, location, qualification, year_of_experience, work_history]
Index: []

In [457]:
finalResumeSecond.shape

(9544, 7)

In [476]:
finalResumeSecond.head(20)

Unnamed: 0                                             skills  \
0            0  ['Hdfs', 'Core Java', 'Hive', 'Talend', 'YARN'...   
1            1  ['R', 'SAS', 'Data Visualization', 'Business A...   
2            2  ['HTML', 'Hive', 'PySpark', 'Application Suppo...   
3            3  ['commercial auto underwriting', 'mortgage loa...   
4            4  ['administrative', 'audit procedures', 'genera...   
5            5  ['Network Administration', 'Cisco ASA firewall...   
6            6  [None, 'Machine Learning', 'Linear Regression'...   
7            7  ['Manufacturing Process', 'First Article Inspe...   
8            8  ['Multivariate Analysis', 'Random Forest', 'Em...   
9            9  ['Machine learning', 'Natural language Process...   
10          10  ['Front Office', 'Finance', 'VideoScribe Softw...   
11          11  ['R', 'SAS', 'Artificial Intelligence', 'SQL',...   
12          12  [None, 'GraphQL', 'Linear Regression', 'Data A...   
13          13  ['HTML', 'PyCharm', 'MS Office', 'OpenCV', 'CS...   
14          14  ['Feature Development', 'HTML', 'React Native'...   
15          15  ['Machine Learning', 'Data Visualization', 'My...   
16          16  ['Operating Systems', 'networking', 'Maintain ...   
17          17  ['Machine Learning', 'OCR', 'Artificial Intell...   
18          18  ['Feature Development', 'HTML', 'React Native'...   
19          19  ['Machine learning', 'XGBoost', 'Classificatio...   

            languages       location  \
0                 NaN            NaN   
1                 NaN            NaN   
2                 NaN            NaN   
3                 NaN            NaN   
4                 NaN            NaN   
5                 NaN            NaN   
6                 NaN            NaN   
7   Spanish (Unknown)            NaN   
8                 NaN            NaN   
9                 NaN            NaN   
10                NaN            NaN   
11                NaN            NaN   
12                NaN            NaN   
13                NaN            NaN   
14                NaN  Sunnyvale, CA   
15                NaN            NaN   
16                NaN            NaN   
17                NaN            NaN   
18                NaN  Sunnyvale, CA   
19                NaN            NaN   

                                        qualification  year_of_experience  \
0   big data analytics working and database wareho...            5.275838   
1   fresher looking to join as a data analyst and ...            5.442847   
2   , Birla Institute of Technology (BIT), Ranchi,...                 NaN   
3   to obtain a position in a fastpaced business o...                 NaN   
4   professional accountant with an outstanding wo...                 NaN   
5   to secure an it specialist desktop support net...                 NaN   
6            , DJR College and University, B.Tech, IT            0.665298   
7   , POLYTECHNIC UNIVERSITY OF PUERTO RICO, Bache...                 NaN   
8   certified data analyst with a degree in electr...            5.694730   
9   , Dr. Virendra Swaroop Institute of Computer S...            6.108145   
10  enthusiastic knowledgeable professional with e...                 NaN   
11                , NARSEE MONJEE INSTITUTE, BCA, N/A            4.859685   
12  , Raja Kashwant Singh General College, B.A, Ec...            5.193703   
13  energetic innovative engineering undergraduate...                 NaN   
14  soontobe computer science graduate eager to co...            3.778234   
15  , CDAC ACTS, B.Tech, Electronics and Telecommu...            4.692676   
16  , University of Phoenix Texas, Business Admini...           20.610541   
17  machine learning engineer seeking assignments ...            5.193703   
18  soontobe computer science graduate eager to co...            3.778234   
19      , Raja Birsimbal College, Ranchi, B.Tech, CSE            5.946612   

                                         work_history  
0   Positions: Big Data Analyst | Respo

In [478]:
for i in range(20):
  print(i, " is ")
  for j in finalResumeSecond.columns:
    print(finalResumeSecond[j][i])
  print("    -----------------------------------------------------------")

0  is 
0
['Hdfs', 'Core Java', 'Hive', 'Talend', 'YARN', 'Microsoft Azure', 'Java', 'Hadoop', 'Data Structures', 'Machine Learning', 'Cloud', 'Amazon Redshift', 'RDBMS', 'Python', 'Informatica', 'Data Science', 'C++', 'Spark', 'Mapreduce', 'DBMS', 'Big Data']
nan
nan
big data analytics working and database warehouse manager with robust experience in handling all kinds of data i have also used multiple cloud infrastructure services and am well acquainted with them currently in search of role that offers more of development, The Amity School of Engineering & Technology (ASET), Noida, B.Tech, Electronics
5.27583846680356
Positions: Big Data Analyst | Responsibilities: Troubleshooting, System Monitoring, Technical Support, Industry Trends, Documentation, Software Deployment, Training & Mentorship, Collaboration, Field Visits
    -----------------------------------------------------------
1  is 
1
['R', 'SAS', 'Data Visualization', 'Business Analytics', 'Machine Learning', 'PowerBi', 'Data 

In [480]:
finalResumeSecond.languages.isna().sum()


8844

In [481]:
finalResumeSecond.location.isna().sum()

8844

In [482]:
Final_Resume_without_languageColumn_locationColumn = finalResumeSecond
Final_Resume_without_Null_ofLanguage_and_location = finalResumeSecond

In [483]:
Final_Resume_without_languageColumn_locationColumn = Final_Resume_without_languageColumn_locationColumn.drop(columns=["languages", "location"])
Final_Resume_without_languageColumn_locationColumn.dropna()
Final_Resume_without_languageColumn_locationColumn.head(), Final_Resume_without_languageColumn_locationColumn.shape

(   Unnamed: 0                                             skills  \
 0           0  ['Hdfs', 'Core Java', 'Hive', 'Talend', 'YARN'...   
 1           1  ['R', 'SAS', 'Data Visualization', 'Business A...   
 2           2  ['HTML', 'Hive', 'PySpark', 'Application Suppo...   
 3           3  ['commercial auto underwriting', 'mortgage loa...   
 4           4  ['administrative', 'audit procedures', 'genera...   
 
                                        qualification  year_of_experience  \
 0  big data analytics working and database wareho...            5.275838   
 1  fresher looking to join as a data analyst and ...            5.442847   
 2  , Birla Institute of Technology (BIT), Ranchi,...                 NaN   
 3  to obtain a position in a fastpaced business o...                 NaN   
 4  professional accountant with an outstanding wo...                 NaN   
 
                                         work_history  
 0  Positions: Big Data Analyst | Responsibilities...  
 1  Posi

In [484]:
Final_Resume_without_languageColumn_locationColumn.to_csv("ResumswithoutLanguage&Location.csv")

In [485]:
Final_Resume_without_Null_ofLanguage_and_location = Final_Resume_without_Null_ofLanguage_and_location.dropna()
Final_Resume_without_Null_ofLanguage_and_location.shape , Final_Resume_without_Null_ofLanguage_and_location.head(10)

((0, 7),
 Empty DataFrame
 Columns: [Unnamed: 0, skills, languages, location, qualification, year_of_experience, work_history]
 Index: [])